src/number_theory/modular_forms/basic.lean

```diff
@@ -0,0 +1,286 @@
+  rw [mul_assoc, h55, ←int.coe_cast_ring_hom, ←matrix.special_linear_group.coe_matrix_coe,
+    matrix.special_linear_group.det_coe ((γ : SL(2, ℤ)) : SL(2, ℝ))],
+  simp only [of_real_one, one_zpow, mul_one], },
+  { intros hf γ,
+  simp_rw [slash_action_eq_slash],
+  ext1,
+  rw [slash, ←upper_half_plane.subgroup_moeb, (hf γ x), mul_comm],
+  have h55 := zpow_neg_mul_zpow_self k (upper_half_plane.denom_ne_zero (γ : GL(2, ℝ)⁺) x),
+  simp_rw upper_half_plane.denom at *,
+  simp only [matrix.special_linear_group.coe_GL_pos_coe_GL_coe_matrix, coe_coe,
+    matrix.special_linear_group.coe_matrix_coe, int.coe_cast_ring_hom, matrix.map_apply,
+    of_real_int_cast] at h55,
+  simp only [coe_coe, matrix.special_linear_group.coe_GL_pos_coe_GL_coe_matrix,
+    matrix.map_apply, of_real_int_cast],
+  rw (matrix.special_linear_group.det_coe ((γ : SL(2, ℤ)) : SL(2, ℝ))),
+  simp only [matrix.special_linear_group.coe_matrix_coe, int.coe_cast_ring_hom, matrix.map_apply,
+    of_real_int_cast, of_real_one, one_zpow, mul_one],
+  simp_rw [← mul_assoc, h55],
+  simp },
+end
+
+lemma mul_modular (k_1 k_2 : ℤ) (Γ : subgroup SL(2,ℤ)) (f g : ℍ → ℂ)
+  (hf : f ∈ weakly_modular_submodule k_1 Γ) (hg : g ∈ weakly_modular_submodule k_2 Γ) :
+  f * g ∈ weakly_modular_submodule (k_1 + k_2) Γ :=
+begin
+  simp only [weakly_modular_mem', pi.mul_apply, coe_coe] at *,
+  intros γ z,
+  rw [(hf γ z), (hg γ z)],
+  have pown := zpow_add₀ (upper_half_plane.denom_ne_zero (γ : GL(2, ℝ)⁺) z) k_1 k_2,
+  simp only [upper_half_plane.denom, coe_fn_coe_base, ne.def,
+    matrix.general_linear_group.coe_fn_eq_coe, coe_coe] at pown,
+  rw pown,
+  ring,
+end
+
+/-- A function `f : ℍ → ℂ` is a modular form weight `k ∈ ℤ` and of level `Γ` if it is holomorphic,
+ weakly modular and bounded at infinity. -/
+structure is_modular_form_of_weight_and_level (k : ℤ) (Γ : subgroup SL(2,ℤ)) (f : ℍ → ℂ) : Prop :=
+(hol : mdifferentiable 𝓘(ℂ) 𝓘(ℂ) (↑f : ℍ' → ℂ))
+(transf : f ∈ weakly_modular_submodule k Γ)
+(infinity : ∀ (A : SL(2,ℤ)), is_bounded_at_im_infty (f ∣[k, A]))
+
+/-- A function `f : ℍ → ℂ` is a cusp form of weight `k ∈ ℤ` and of level `Γ` if it is holomorphic,
+ weakly modular, and zero at infinity. -/
+structure is_cusp_form_of_weight_and_level (k : ℤ) (Γ : subgroup SL(2,ℤ)) (f : ℍ → ℂ) : Prop :=
+(hol : mdifferentiable 𝓘(ℂ) 𝓘(ℂ) (↑f : ℍ' → ℂ))
+(transf : f ∈ weakly_modular_submodule k Γ)
+(infinity : ∀ (A : SL(2,ℤ)), is_zero_at_im_infty (f ∣[k, A]))
+
+/-- The zero modular form is a cusp form-/
+lemma zero_cusp_form : is_cusp_form_of_weight_and_level k Γ 0 :=
+{ hol := by {apply mdifferentiable_zero,},
+  transf := (weakly_modular_submodule k Γ).zero_mem',
+  infinity := by { intro A,
+    rw slash_action.mul_zero,
+    apply (zero_at_im_infty_submodule ℂ).zero_mem}}
+
+lemma is_modular_form_of_weight_and_level_of_is_cusp_form_of_weight_and_level {f : ℍ → ℂ}
+  (h : is_cusp_form_of_weight_and_level k Γ f) : is_modular_form_of_weight_and_level k Γ f :=
+{ hol := h.1,
```

Comment:
```suggestion
{ hol := h.hol,
```
Here and throughout -- the proof is much more readable if you use the "word" version of the projection, rather than the numeric version.